### DISCUSSION TITLE
Incorporate Polygon Triangulation from GEOS

### DISCUSSION THREAD

**Discussion by mapcreatorGuus** [shapely/shapely#1748] [Ideas]
I ran into issues triangulating polygons because the triangulate function simply uses all the points of the polygon instead of the polygon itself. In more complex polygons, in particular those with holes, this quickly leads to situations where certain triangles cross the original geometry. This is in accordance with the documentation and the way the triangulate function in GEOS is specified. See also these posts:
https://gis.stackexchange.com/questions/316697/delaunay-triangulation-algorithm-in-shapely-producing-erratic-result
https://stackoverflow.com/questions/65019170/how-do-you-triangulate-a-polygon-in-shapely?noredirect=1#comment125015008_65019170
There are workarounds, but they open up awkward edge-cases, and come with a performance cost: python will never be as fast as a GEOS binding.

Luckily for us, GEOS has added the exact functionality I need in version 3.10:
https://libgeos.org/doxygen/classgeos_1_1triangulate_1_1polygon_1_1ConstrainedDelaunayTriangulator.html
https://libgeos.org/posts/2021-10-01-geos-3-10-released/

I think it would be very nice for shapely to incorporate this function (and of course, many more of the latest GEOS functions). In my opinion, the closer shapely and GEOS are functionality-wise, the better we can enjoy the easy-of-use and flexibility of the Python language while performing complicated geometrical functions efficiently.

Let me know what you think.

PS. In case anyone is interested, while this is not yet there, I will try to update GEOS in PostGIS and perform the triangulation on the database side.

**Comment by jorisvandenbossche** [comment]
Yes, we certainly intend to include this new GEOS feature in Shapely. The issue https://github.com/shapely/shapely/issues/1570 is keeping track of all new GEOS features that have not yet been exposed in Shapely. Specifically for `GEOSConstrainedDelaunayTriangulation`, there is actually an open PR (https://github.com/shapely/shapely/pull/1685)
